### PR TITLE
Move generic information from decorated pointers to decorated tags

### PR DESCRIPTION
**Pull request by smowton**
`java_generic_parametert` is reimplemented as a plain pointer to a decorated `struct_tag_typet` (currently always a `java.lang.Object`, but that will change when we support bounded generics). `java_generic_typet` is reimplemented as a plain pointer to the existing type `java_generic_struct_tag_typet`, which already expressed nested superclasses and now also specifies references.

The existing types (subclasses of `java_reference_typet`) are retained, but do not store any data themselves and simply forward their accessors to the underlying tag, meaning no changes are needed to existing code.

This means that a pointer is once again just a pointer, and can be removed and reapplied
without risking losing information. On the other hand tags now carry decorations more often
(they already played this role when expressing a superclass, which is implemented as a nested
struct), so `ns.follow` returns a type missing its decorations -- however this is hopefully
a small risk, as code using the decorated-pointer implementation would already need to know
that following the pointer, never mind the tag at the end of it, could lose information.

### CODE REVIEW

**Comment by smowton** [comment]
I've prototyped this with test-gen btw and seems all tests pass with no changes to its code

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4195069f2985640f3f8dc0a196cb58de0502563d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100136027

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: dd3afa19d0b93c75be139ff27b03f0e1ba872f66).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100212017

**Review by tautschnig — APPROVED** [review]
🎉 for not having decorated pointers!

File: src/util/irep_ids.def [hunk 1]
```diff
@@ -648,11 +648,10 @@ IREP_ID_ONE(switch_case_number)
 IREP_ID_ONE(java_array_access)
 IREP_ID_ONE(java_member_access)
 IREP_ID_TWO(C_java_generic_parameter, #java_generic_parameter)
-IREP_ID_TWO(C_java_generic_type, #java_generic_type)
 IREP_ID_TWO(C_java_generics_class_type, #java_generics_class_type)
 IREP_ID_TWO(C_java_implicitly_generic_class_type, #java_implicitly_generic_class_type)
 IREP_ID_TWO(C_java_generic_symbol, #java_generic_symbol)
-IREP_ID_TWO(generic_types, #generic_types)
+IREP_ID_TWO(generic_types, generic_types)
```
Comment (owen-mc-diffblue) [inline src/util/irep_ids.def]: Can this be `IREP_ID_ONE` now?

**Comment by smowton** [comment]
@kroening can I take your thumbs-up for an approval?

**Comment by tautschnig** [comment]
@smowton clang-format seems unhappy.

**Comment by smowton** [comment]
It appears clang-format was flagging stuff that had nothing to do with this PR? I've rebased to see if that fixes things, will merge if/when CI passes.

**Comment by tautschnig** [comment]
> It appears clang-format was flagging stuff that had nothing to do with this PR?

I've seen that quite a lot already, but admittedly never investigated properly. It does seem to happen when there are a lot of PRs in flight.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 68878636c5d083911325cce0ab8e46689242b4c3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100429555

**Review by thk123 — APPROVED** [review]

**Pull request merged**